pkg/chunk/chunk_store.go

```diff
@@ -482,3 +482,109 @@ func (c *store) convertChunkIDsToChunks(ctx context.Context, userID string, chun
 
 	return chunkSet, nil
 }
+
+func (c *store) DeleteChunk(ctx context.Context, from, through model.Time, userID, chunkID string, metric labels.Labels, partiallyDeletedInterval *model.Interval) error {
+	metricName := metric.Get(model.MetricNameLabel)
+	if metricName == "" {
+		return errors.New("No metric name label")
+	}
+
+	chunkWriteEntries, err := c.schema.GetWriteEntries(from, through, userID, string(metricName), metric, chunkID)
+	if err != nil {
+		return err
+	}
+
+	return c.deleteChunk(ctx, userID, chunkID, metric, chunkWriteEntries, partiallyDeletedInterval, func(chunk Chunk) error {
+		return c.PutOne(ctx, chunk.From, chunk.Through, chunk)
+	})
+}
+
+func (c *store) deleteChunk(ctx context.Context, userID, chunkID string, metric labels.Labels,
+	chunkWriteEntries []IndexEntry, partiallyDeletedInterval *model.Interval, putChunkFunc func(chunk Chunk) error) error {
+	metricName := metric.Get(model.MetricNameLabel)
+	if metricName == "" {
+		return errors.New("No metric name label")
+	}
+
+	// if chunk is partially deleted, fetch it, slice non-deleted portion and put it to store before deleting original chunk
+	if partiallyDeletedInterval != nil {
+		err := c.deleteChunkPartially(ctx, userID, chunkID, *partiallyDeletedInterval, putChunkFunc)
+		if err != nil {
+			return err
+		}
+	}
+
+	batch := c.index.NewWriteBatch()
+	for i := range chunkWriteEntries {
+		batch.Delete(chunkWriteEntries[i].TableName, chunkWriteEntries[i].HashValue, chunkWriteEntries[i].RangeValue)
+	}
+
+	err := c.index.BatchWrite(ctx, batch)
+	if err != nil {
+		return err
+	}
+
+	return c.chunks.DeleteChunk(ctx, chunkID)
```

Comment:
We should check the error and ignore it (do not propagate it) if it's `chunk.ErrStorageObjectNotFound`. It's better to do the check here because the error could be wrapped to add more contextual information in parent callers.

Reply:
I wanted to leave it up to the caller to decide whether its a problem or not. I will give it a thought.